ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/damle/SandboxDamle.scala

```diff
@@ -36,7 +41,23 @@ object SandboxDamle {
     def resolveStep(result: Result[A]): Future[Either[DamlLfError, A]] = {
       result match {
         case ResultNeedPackage(packageId, resume) =>
-          getPackage(packageId).flatMap(mbPkg => resolveStep(resume(mbPkg)))
+          var gettingPackage = false
+          packagePromises
+            .computeIfAbsent(packageId, { _ =>
+              val p = Promise[Option[Package]]()
+              gettingPackage = true
+              getPackage(packageId).foreach(p.success)
+              p
+            })
+            .future
+            .flatMap { mbPkg =>
+              if (gettingPackage && mbPkg.isEmpty) {
+                // Failed to find the package. Remove the promise to allow later retries.
```

Comment:
This doesn't include when the future is failed (e.g. db connection failed this time).

Reply:
Thanks, I was sure I missed something. Suggestions how to handle that? Or suggestions how to write this whole thing in a safer way?